xunit-autolabeler-v2/ast_parser/core/analyze.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from os import path
+from typing import List, Tuple
+
+from . import polyglot_drift_data as pdd, polyglot_parser, yaml_utils
+
+
+def analyze_json(
+    repo_json: str,
+    root_dir: str
+) -> Tuple[List[str], List[str], List[str], List[pdd.PolyglotDriftData]]:
+    """Perform language-agnostic AST analysis on a directory
+
+    This function processes a given directory's language-specific
+    analysis (stored in a polyglot_snippet_data.json file) into a
+    list of automatically detected snippets. It then augments the
+    automatic detection results with useful manual data (specified
+    in .drift-data.yml files). Finally, it repackages all this data
+    into a tuple containing 4 useful lists of data as shown in the
+    'returns' section.
+
+    Arguments:
+        repo_json: A path to a polyglot_snippet_data.json file
+                   generated for the specified root_dir
+        root_dir: The root directory to perform AST analysis on
+
+    Returns:
+        A tuple containing the following:
+         - A list of tags found (via grep/text search)
+           within the given directory and its subdirectories
+         - A list of tags detected (by the AST parser)
+           within the given directory and its subdirectories
+         - A list of tags that the AST parser detected,
+           but chose to ignore (due to constants or user
+           specification in .drift-data.yml files)
+         - A list of snippet objects (as typed NamedTuples)
+           detected by the AST parser in the given directory
+           and its subdirectories
+    """
+    tuple_methods = []
```

Comment:
Why an array and not a set?

Reply:
We're changing (_not_ modifying^) the elements in this list. IMO it's easier to do this via `enumerate()` than adding to/removing from a `set`.

^: the elements in this list are of type `PolyglotDriftData`, which is a subclass of `NamedTuple`. `NamedTuple` objects are **not** mutable. ([There are alternatives](https://stackoverflow.com/questions/29290359/existence-of-mutable-named-tuple-in-python), but I'm not sure whether it's worth switching to them.)

Reply:
If this are meant to be mutated then yes, recordclass could be a good alternative here.

Reply:
Cool - will switch to those. (I can probably clean up the `enumerate`s a bit too. 🙂 )